audformat/core/utils.py

```diff
@@ -36,66 +60,108 @@ def concat(
-        ...     index=filewise_index(['f1', 'f2', 'f3']),
-        ...     name='number',
+        >>> obj1 = pd.Series(
+        ...     [0., 1.],
+        ...     index=filewise_index(['f1', 'f2']),
+        ...     name='float',
         ... )
-        >>> df = pd.DataFrame(
+        >>> obj2 = pd.DataFrame(
         ...     {
-        ...         'string': ['a', 'b', 'c'],
+        ...         'float': [1., 2.],
+        ...         'string': ['a', 'b'],
         ...     },
-        ...     index=filewise_index(['f2', 'f3', 'f4']),
+        ...     index=segmented_index(['f2', 'f3']),
         ... )
-        >>> concat([series, df])
-              number string
-        file
-        f1       1.0    NaN
-        f2       2.0      a
-        f3       3.0      b
-        f4       NaN      c
+        >>> concat([obj1, obj2])
+                         float string
+        file start  end
+        f1   0 days NaT    0.0    NaN
+        f2   0 days NaT    1.0      a
+        f3   0 days NaT    2.0      b
 
     """
     if not objs:
-        return pd.DataFrame([], index=filewise_index())
+        return pd.Series([], index=filewise_index(), dtype='object')
 
-    concat_table_type = define.IndexType.FILEWISE
-    for frame in objs:
-        if index_type(frame) == define.IndexType.SEGMENTED:
-            concat_table_type = define.IndexType.SEGMENTED
-
-    if concat_table_type == define.IndexType.SEGMENTED:
-        index = segmented_index()
-        objs = [to_segmented_index(frame) for frame in objs]
-    else:
-        index = filewise_index()
+    # the new index is a union of the individual objects
+    index = union([obj.index for obj in objs])
+    as_segmented = index_type(index) == define.IndexType.SEGMENTED
 
-    index = index.append([frame.index for frame in objs])
-    index = index.drop_duplicates()
-
-    columns = OrderedDict()
+    # list with all columns we need to concatenate
+    columns = []
     for obj in objs:
         if isinstance(obj, pd.Series):
-            columns[obj.name] = obj.dtype
+            columns.append(obj)
         else:
-            for c, d in zip(obj.columns, obj.dtypes):
-                if c not in columns:
-                    columns[c] = d
-
-    df_concat = pd.DataFrame(index=index, columns=columns.keys()).sort_index()
-    for obj in objs:
-        if not obj.empty:
-            if isinstance(obj, pd.Series):
-                df_concat.loc[obj.index, obj.name] = obj
-            else:
-                df_concat.loc[obj.index, obj.columns] = obj
+            for column in obj:
+                columns.append(obj[column])
+
+    # reindex all columns to the new index
+    columns_reindex = OrderedDict()
+    for column in columns:
+
+        if as_segmented:
+            column = to_segmented_index(column)
+
+        # if we already have a column with that name, we have to merge them
+        if column.name in columns_reindex:
+
+            # assert same dtype
+            if not compare_dtype(
```

Comment:
`compare_dtype` seems not to be the best name here, because it's not obvious what its return value should be. Maybe rename to something like `same_dtype` or `have_same_dtype`?

Reply:
changed to `same_dtype()`